File: spawn-ssh-suid.c
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <sys/un.h>
#include <errno.h>
#include <pwd.h>

#define USER "www-data"

int main(int argc, char *argv[]) {
    int fd, r;
    socklen_t l;
    union {
            struct sockaddr sa;
            struct sockaddr_un un;
            struct sockaddr_in in;
            struct sockaddr_in6 in6;
            struct sockaddr_storage storage;
    } local, remote;
    pid_t ppid;
    struct stat st;
    struct passwd *pw;

    errno = 0;
    pw = getpwnam(USER);
    if (errno > 0 || !pw)
        goto fail;

    if (pw->pw_uid != getuid())
        goto fail;

    fd = open("/dev/null", 0);
    if (fd < 0)
        goto fail;

    if (dup2(fd, STDERR_FILENO) < 0)
        goto fail;

    if (clearenv() < 0)
        goto fail;

    fd = 3;
    l = sizeof(local);
    if (getsockname(fd, &local.sa, &l) < 0)
        goto fail;

    l = sizeof(remote);
    if (getpeername(fd, &remote.sa, &l) < 0)
        goto fail;

    switch (local.sa.sa_family) {
    case AF_INET:
    case AF_INET6:
    case AF_UNIX:
        break;
    default:
        goto fail;
    }

    if (dup2(fd, STDOUT_FILENO) < 0)
        goto fail;
    if (dup2(fd, STDIN_FILENO) < 0)
        goto fail;
    r = close(fd);
    if (r < 0 && errno != EINTR)
        goto fail;

    execve("/usr/sbin/sshd", (char *[]){"sshd", "-i", NULL}, (char *[]){NULL});
fail:
    return EXIT_FAILURE;
}

File: spawn-ssh-suid.go
package main

import (
	"os"
	"os/user"
	"strconv"
	"syscall"
)

const (
	from_user string = "www-data"
	fd        int    = 3
)

func main() {
	var e error
	u := new(user.User)
	u, e = user.Lookup(from_user)
	if e != nil {
		os.Exit(1)
	}

	uid, e := strconv.Atoi(u.Uid)
	if e != nil {
		os.Exit(1)
	}

	if os.Getuid() != uid {
		os.Exit(1)
	}

	// doesn't work cause syscall.Sockaddr type
	// can't be used in any way.
	//syscall.Getsockaddr(fd)
	e = syscall.Dup2(fd, syscall.Stdin)
	if e != nil {
		os.Exit(1)
	}
	e = syscall.Dup2(fd, syscall.Stdout)
	if e != nil {
		os.Exit(1)
	}
	e = syscall.Dup2(fd, syscall.Stderr)
	if e != nil {
		os.Exit(1)
	}
	e = syscall.Close(fd)
	if e != nil {
		os.Exit(1)
	}

	syscall.Setuid(0)
	_ = syscall.Exec("/usr/sbin/sshd", []string{"sshd", "-i"}, []string{})
	os.Exit(1)
}
